libgadget/uvbg.c

```diff
@@ -0,0 +1,783 @@
+        memcpy(sfr_filtered, sfr_unfiltered, sizeof(fftwf_complex) * slab_n_complex);
+
+        // do the filtering unless this is the last filter step
+        int local_ix_start = (int)(UVBGgrids.slab_ix_start[this_rank]);
+        if (!flag_last_filter_step) {
+            filter(deltax_filtered, local_ix_start, local_nix, UVBG_DIM, (float)R);
+            filter(stars_slab_filtered, local_ix_start, local_nix, UVBG_DIM, (float)R);
+            filter(sfr_filtered, local_ix_start, local_nix, UVBG_DIM, (float)R);
+        }
+
+        // inverse fourier transform back to real space
+        fftwf_execute_dft_c2r(UVBGgrids.plan_dft_c2r, deltax_filtered, (float*)deltax_filtered);
+        fftwf_execute_dft_c2r(UVBGgrids.plan_dft_c2r, stars_slab_filtered, (float*)stars_slab_filtered);
+
+        // Perform sanity checks to account for aliasing effects
+        for (int ix = 0; ix < local_nix; ix++)
+            for (int iy = 0; iy < UVBG_DIM; iy++)
+                for (int iz = 0; iz < UVBG_DIM; iz++) {
+                    i_padded = grid_index(ix, iy, iz, UVBG_DIM, INDEX_PADDED);
+                    ((float*)deltax_filtered)[i_padded] = fmaxf(((float*)deltax_filtered)[i_padded], -1 + FLOAT_REL_TOL);
+                    ((float*)stars_slab_filtered)[i_padded] = fmaxf(((float*)stars_slab_filtered)[i_padded], 0.0f);
+                    ((float*)sfr_filtered)[i_padded] = fmaxf(((float*)sfr_filtered)[i_padded], -1 + FLOAT_REL_TOL);
+                }
+
+
+        // ============================================================================================================
+        // {
+        //     // DEBUG HERE
+        //     const int grid_size = (int)(local_nix * UVBG_DIM * UVBG_DIM);
+        //     float* grid = (float*)calloc(grid_size, sizeof(float));
+        //     int count_gtz = 0;
+        //     for (int ii = 0; ii < local_nix; ii++)
+        //         for (int jj = 0; jj < UVBG_DIM; jj++)
+        //             for (int kk = 0; kk < UVBG_DIM; kk++) {
+        //                 grid[grid_index(ii, jj, kk, UVBG_DIM, INDEX_REAL)] = ((float*)deltax_filtered)[grid_index(ii, jj, kk, UVBG_DIM, INDEX_PADDED)];
+        //                 if (grid[grid_index(ii, jj, kk, UVBG_DIM, INDEX_REAL)] > 0)
+        //                     count_gtz++;
+        //             }
+
+        //     message(0, "count_gtz for filter R=%.2f = %d\n", R, count_gtz);
+
+        //     BigFile fout;
+        //     char fname[256];
+        //     sprintf(fname, "output/filterstep-%.2f.bf", R);
+        //     big_file_mpi_create(&fout, fname, MPI_COMM_WORLD);
+        //     BigBlock block;
+        //     int n_ranks;
+        //     MPI_Comm_size(MPI_COMM_WORLD, &n_ranks);
+        //     big_file_mpi_create_block(&fout, &block, "deltax", "=f4", 1, n_ranks, UVBG_DIM*UVBG_DIM*UVBG_DIM, MPI_COMM_WORLD);
+        //     BigBlockPtr ptr = {0};
+        //     int start_elem = this_rank > 1 ? UVBGgrids.slab_nix[this_rank - 1]*UVBG_DIM*UVBG_DIM : 0;
```

Comment:
I don't think you want this seek. mpi_write assumes you want to start writing from the BigBlockPtr of the 0 rank, then assumes data on ranks is contiguous.